lib/evaluate/host.h

```diff
@@ -0,0 +1,229 @@
+        CastFortranToHost<typename FTN_T::Part>(x.AIMAG())};
+  } else {
+    return *reinterpret_cast<const HostType<FTN_T> *>(&x);
+  }
+}
+
+template<typename T> struct BiggerOrSameHostTypeHelper {
+  using Type =
+      std::conditional_t<HostTypeExists<T>(), HostType<T>, UnsupportedType>;
+  using FortranType = T;
+};
+
+template<typename FTN_T>
+using BiggerOrSameHostType = typename BiggerOrSameHostTypeHelper<FTN_T>::Type;
+template<typename FTN_T>
+using BiggerOrSameFortranTypeSupportedOnHost =
+    typename BiggerOrSameHostTypeHelper<FTN_T>::FortranType;
+
+template<typename... T> constexpr inline bool BiggerOrSameHostTypeExists() {
+  return (... && (!std::is_same_v<BiggerOrSameHostType<T>, UnsupportedType>));
+}
+
+// Defining the actual mapping
+template<> struct HostTypeHelper<Type<TypeCategory::Integer, 1>> {
+  using Type = std::int8_t;
+};
+
+template<> struct HostTypeHelper<Type<TypeCategory::Integer, 2>> {
+  using Type = std::int16_t;
+};
+
+template<> struct HostTypeHelper<Type<TypeCategory::Integer, 4>> {
+  using Type = std::int32_t;
+};
+
+template<> struct HostTypeHelper<Type<TypeCategory::Integer, 8>> {
+  using Type = std::int64_t;
+};
+
+// no int 128bit
+
+// no float 16bits
+
+template<> struct HostTypeHelper<Type<TypeCategory::Real, 4>> {
+  // IEE 754 64bits
+  using Type = std::conditional_t<sizeof(float) == 4 &&
+          std::numeric_limits<float>::is_iec559,
+      float, UnsupportedType>;
+};
+
+template<> struct HostTypeHelper<Type<TypeCategory::Real, 8>> {
+  // IEE 754 64bits
+  using Type = std::conditional_t<sizeof(double) == 8 &&
+          std::numeric_limits<double>::is_iec559,
+      double, UnsupportedType>;
+};
+
+template<> struct HostTypeHelper<Type<TypeCategory::Real, 10>> {
+  // X87 80bits
+  using Type = std::conditional_t<sizeof(long double) >= 10 &&
+          std::numeric_limits<long double>::digits == 64 &&
+          std::numeric_limits<long double>::max_exponent == 16384,
+      long double, UnsupportedType>;
+};
+
+template<> struct HostTypeHelper<Type<TypeCategory::Real, 16>> {
+  // IEE 754 128bits
+  using Type = std::conditional_t<sizeof(long double) == 16 &&
+          std::numeric_limits<long double>::digits == 113 &&
+          std::numeric_limits<long double>::max_exponent == 16384,
+      long double, UnsupportedType>;
+};
+
+template<int KIND> struct HostTypeHelper<Type<TypeCategory::Complex, KIND>> {
+  using RealT = Fortran::evaluate::Type<TypeCategory::Real, KIND>;
+  using Type = std::conditional_t<HostTypeExists<RealT>(),
+      std::complex<HostType<RealT>>, UnsupportedType>;
+};
+
+template<int KIND> struct HostTypeHelper<Type<TypeCategory::Character, KIND>> {
+  using Type = std::conditional_t<KIND <= 8, std::uint8_t, UnsupportedType>;
+};
+
+template<> struct HostTypeHelper<Type<TypeCategory::Character, 1>> {
+  using Type = std::string;
+};
+
+template<> struct HostTypeHelper<Type<TypeCategory::Character, 2>> {
+  using Type = std::u16string;
+};
+
```

Comment:
Missing `CHARACTER(KIND=4)`.

You could use the typedefs in `type.h` for these types of string constants rather than duplicating them.

Reply:
Thanks, I missed it. that will make maintenance easier.